src/main/java/org/carlmontrobotics/subsystems/Limelight.java

```diff
@@ -0,0 +1,81 @@
+package org.carlmontrobotics.subsystems;
+
+import static org.carlmontrobotics.Constants.Limelight.*;
+import static org.carlmontrobotics.Constants.Limelight.Apriltag.*;
+
+
+import edu.wpi.first.math.MathUtil;
+import edu.wpi.first.math.estimator.SwerveDrivePoseEstimator;
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Pose3d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.util.Units;
+import edu.wpi.first.networktables.NetworkTable;
+import edu.wpi.first.networktables.NetworkTableEntry;
+import edu.wpi.first.networktables.NetworkTableInstance;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+
+public class Limelight extends SubsystemBase {
+  private final Drivetrain drivetrain;
+  private final SwerveDrivePoseEstimator poseEstimator;
+  private double tv, tx;
+  private double[] botPose = null;
+  private double[] targetPose = null;
+  private Pose3d botpose;
+
+  //private double distOffset, horizOffset;
+  //private double horizHeadingError, horizAdjust;
+
+  public Limelight(Drivetrain drivetrain) {
+    this.drivetrain = drivetrain;
+    poseEstimator = new SwerveDrivePoseEstimator(
+      drivetrain.getKinematics(), 
+      Rotation2d.fromDegrees(drivetrain.getHeading()), 
+      drivetrain.getModulePositions(), 
+      new Pose2d());
+    
+    botPose = NetworkTableInstance.getDefault().getTable(shooterLimelightName).getEntry("botpose").getDoubleArray(new double[7]);
+    //targetPose = table.getEntry("targetpose_fieldspace").getDoubleArray(new double[7]);
+  }
+
+  @Override
+  public void periodic() {
+    getDistanceToTargetSpeaker();
+    getCurrentPose();
+  }
+
+  public void updateBotPose3d(){
+    botpose = LimelightHelpers.getBotPose3d(shooterLimelightName);
+  }
+
+
+  public Pose2d getCurrentPose(){
+    System.out.println(poseEstimator.getEstimatedPosition());
+    return poseEstimator.getEstimatedPosition();
+  }
+
+
+  public double getDistanceToTargetSpeaker(){
+    if (LimelightHelpers.getFiducialID(shooterLimelightName) == SPEAKER_CENTER_TAG_ID_1 || LimelightHelpers.getFiducialID("limelight") == SPEAKER_CENTER_TAG_ID_2){
```

Comment:
Your math assumes that the robot is facing the apriltag perfectly or in other words, `tx` is 0. Unless you want to account for tx, wouldn't you want to be facing the apriltag in the center of the speaker instead of the side apriltag to determine the correct distance to shoot at?

Reply:
I don't understand what you mean, my code assumes that we are facing the apriltag in the center of the speaker (ids 4 and 7).

Reply:
whoops i thought you were facing the center of the speaker + the other apriltag by the side. I didn't realize it was for red alliance center speaker and blue alliance center speaker. I would suggest that you rename your `SPEAKER_CENTER_TAG_ID_1` to something like `RED_SPEAKER_CENTER_TAG_ID` to make the distinction more clear